File: recipes-frontend/src/app/models/recipe.ts
import { Ingredient } from "./ingredient";

export interface Recipe {
    id: string | undefined;
    name: string;
    instructions: string,
    ingredients: Ingredient[];
}

export function copyOf(recipe: Recipe): Recipe {
    let copy: Recipe = {
        id: recipe.id,
        name: recipe.name,
        instructions: recipe.instructions,
        ingredients: []
    }
    recipe.ingredients.forEach(ingredient => {
        copy.ingredients.push({
            name: ingredient.name, 
            quantity: {
                value: ingredient.quantity.value,
                unit: ingredient.quantity.unit
            }});
    });
    return copy;
}

File: recipes-api/src/main/java/recipes/recipes/api/domain/model/Recipe.java
package recipes.recipes.api.domain.model;

import lombok.Data;

import java.util.List;
import java.util.UUID;

/**
 * Recipe model - aggregate root.
 */
@Data
public class Recipe {
    private final UUID id;
    private final String name;
    private final String instructions;
    private final long servings;
    private final List<Ingredient> ingredients;
}
